### ISSUE TITLE
session.execute() and connection.execute() behaves differently when the SQL query has %  (arithmetic operator) in it.

### ISSUE THREAD

**Issue by srinaveendesu** [sqlalchemy/sqlalchemy#5822]
**Describe the bug**
I am trying to execute the following SQL query `SELECT 1 % 2` against the Snowflake db. The query works fine when executing with a connection object but fails when executing with a session object.

**Expected behavior**
The output of the program would be -> 
`(1,)
(1,)`

**To Reproduce**
1) Modify the code to include username, db_password and db_host
2) Run the below code against a Snowflake DB and the error as below will be seen

**Please do not use Flask-SQLAlchemy or any other third-party extensions or dependencies in test cases**.  The test case must illustrate the problem **without** using any third party SQLAlchemy extensions.  Otherwise, please report the bug to those projects first.

```py
# Insert code here
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker, scoped_session

user_name = ''
db_password = ''
db_host = ''

engine = create_engine(
    'snowflake://{user}:{password}@{host}'.format(
        user=user_name,
        password=db_password,
        host=db_host
    ))
connection = engine.connect()
result = connection.execute('select 1 % 2').fetchone()
print(result)


Session = sessionmaker(bind=engine)
session = scoped_session(Session)
result = session.execute('select 1 % 2').fetchone()
print(result)

```

**Error**

```
# Copy error here. Please include the full stack trace.
(1,)
Traceback (most recent call last):
  File "C:\Users\u1072232\PycharmProjects\pythonProject\myenv\lib\site-packages\sqlalchemy\engine\base.py", line 1283, in _execute_context
    self.dialect.do_execute(
  File "C:\Users\u1072232\PycharmProjects\pythonProject\myenv\lib\site-packages\sqlalchemy\engine\default.py", line 590, in do_execute
    cursor.execute(statement, parameters)
  File "C:\Users\u1072232\PycharmProjects\pythonProject\myenv\lib\site-packages\snowflake\connector\cursor.py", line 601, in execute
    Error.errorhandler_wrapper(self.connection, self,
  File "C:\Users\u1072232\PycharmProjects\pythonProject\myenv\lib\site-packages\snowflake\connector\errors.py", line 97, in errorhandler_wrapper
    cursor.errorhandler(connection, cursor, errorclass, errorvalue)
  File "C:\Users\u1072232\PycharmProjects\pythonProject\myenv\lib\site-packages\snowflake\connector\errors.py", line 68, in default_errorhandler
    raise errorclass(
snowflake.connector.errors.ProgrammingError: 001003 (42000): SQL compilation error:
syntax error line 1 at position 10 unexpected '%'.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:/Users/u1072232/PycharmProjects/pythonProject/bug_raise.py", line 21, in <module>
    result = session.execute('select 1 % 2').fetchone()
  File "C:\Users\u1072232\PycharmProjects\pythonProject\myenv\lib\site-packages\sqlalchemy\orm\scoping.py", line 163, in do
    return getattr(self.registry(), name)(*args, **kwargs)
  File "C:\Users\u1072232\PycharmProjects\pythonProject\myenv\lib\site-packages\sqlalchemy\orm\session.py", line 1291, in execute
    return self._connection_for_bind(bind, close_with_result=True).execute(
  File "C:\Users\u1072232\PycharmProjects\pythonProject\myenv\lib\site-packages\sqlalchemy\engine\base.py", line 1020, in execute
    return meth(self, multiparams, params)
  File "C:\Users\u1072232\PycharmProjects\pythonProject\myenv\lib\site-packages\sqlalchemy\sql\elements.py", line 298, in _execute_on_connection
    return connection._execute_clauseelement(self, multiparams, params)
  File "C:\Users\u1072232\PycharmProjects\pythonProject\myenv\lib\site-packages\sqlalchemy\engine\base.py", line 1133, in _execute_clauseelement
    ret = self._execute_context(
  File "C:\Users\u1072232\PycharmProjects\pythonProject\myenv\lib\site-packages\sqlalchemy\engine\base.py", line 1323, in _execute_context
    self._handle_dbapi_exception(
  File "C:\Users\u1072232\PycharmProjects\pythonProject\myenv\lib\site-packages\sqlalchemy\engine\base.py", line 1517, in _handle_dbapi_exception
    util.raise_(
  File "C:\Users\u1072232\PycharmProjects\pythonProject\myenv\lib\site-packages\sqlalchemy\util\compat.py", line 178, in raise_
    raise exception
  File "C:\Users\u1072232\PycharmProjects\pythonProject\myenv\lib\site-packages\sqlalchemy\engine\base.py", line 1283, in _execute_context
    self.dialect.do_execute(
  File "C:\Users\u1072232\PycharmProjects\pythonProject\myenv\lib\site-packages\sqlalchemy\engine\default.py", line 590, in do_execute
    cursor.execute(statement, parameters)
  File "C:\Users\u1072232\PycharmProjects\pythonProject\myenv\lib\site-packages\snowflake\connector\cursor.py", line 601, in execute
    Error.errorhandler_wrapper(self.connection, self,
  File "C:\Users\u1072232\PycharmProjects\pythonProject\myenv\lib\site-packages\snowflake\connector\errors.py", line 97, in errorhandler_wrapper
    cursor.errorhandler(connection, cursor, errorclass, errorvalue)
  File "C:\Users\u1072232\PycharmProjects\pythonProject\myenv\lib\site-packages\snowflake\connector\errors.py", line 68, in default_errorhandler
    raise errorclass(
sqlalchemy.exc.ProgrammingError: (snowflake.connector.errors.ProgrammingError) 001003 (42000): SQL compilation error:
syntax error line 1 at position 10 unexpected '%'.
[SQL: select 1 %% 2]
(Background on this error at: http://sqlalche.me/e/f405)

```

**Versions.**
 - OS: Windoes
 - Python: 3.8
 - SQLAlchemy: 1.3.17
 - Database: Snowflake Version : 4.43.2
 - DBAPI:

**Additional context**

**Have a nice day!**

**Comment by gordthompson** [comment]
Instead of

``` python
.execute('select 1 % 2')
```

try adding `from sqlalchemy import text` and then see if

``` python
.execute(text('select 1 % 2'))
```

does the trick for you.

**Comment by CaselIT** [comment]
Hi,

The difference you are seeing is because `session.execute` will transform the sql string using `text`, quite `connection.execute` does not. Note that using a string directly with `connection.execute` is deprecated and will raise an warning starting in 1.4 and will eventually behave the same as the session, trasforming the string using `text`, in v2.

Regarding your error, my guess it that the snowflake extension does use a param style that makes the text excapt `%` with a `%%`. You should probably check the extension documentation to see how it does handle it or raise an issue with them

**Comment by srinaveendesu** [comment]
@gordthompson  

I have used `text(' select 1 % 2')` in my code. 
However, I see the same error.

**Comment by zzzeek** [comment]
this sounds like an issue in the snowflake driver, which is doubling up percent signs inappropriately.

When a compiler doubles up a percent sign, that indicates that the driver is expecting "pyformat" or "format" bound parameters, that is, "%(name)s" or "%s" indicates a bound parameter.  in this type of driver, using a percent sign by itself has to be doubled up, %%.   however your snowflake driver is rejecting this, suggesting one of two things:   1. the driver should not be doubling up percent signs  or 2. the driver is *conditionally* making use of pyformat/format parameters but only if parameters are present.   in the latter case, the snowflake SQLAlchemy dialect should most likely ensure that whether or not parameters are passed, the driver still gets escaped percent signs, so that there is no need to change the compiled form of the statement whether or not parameters are present.   it can also try to guess if percents should be escaped based on parameters present or not but this is more difficult to do correctly.  

basically though you want to open an issue at https://github.com/snowflakedb/snowflake-sqlalchemy .

**Comment by zzzeek** [comment]
this seems like an external driver issue.
